lightning/src/routing/scorer.rs

```diff
@@ -44,38 +48,237 @@
 
 use routing;
 
+use ln::msgs::DecodeError;
 use routing::network_graph::NodeId;
+use routing::router::RouteHop;
+use util::ser::{Readable, Writeable, Writer};
+
+use prelude::*;
+use core::ops::Sub;
+use core::time::Duration;
+use io::{self, Read};
 
 /// [`routing::Score`] implementation that provides reasonable default behavior.
 ///
 /// Used to apply a fixed penalty to each channel, thus avoiding long paths when shorter paths with
-/// slightly higher fees are available.
+/// slightly higher fees are available. May also further penalize failed channels.
 ///
 /// See [module-level documentation] for usage.
 ///
 /// [module-level documentation]: crate::routing::scorer
-pub struct Scorer {
-	base_penalty_msat: u64,
+pub type Scorer = ScorerUsingClock::<DefaultClock>;
+
+/// Clock used by [`Scorer`].
+#[cfg(not(feature = "no-std"))]
+pub type DefaultClock = std::time::Instant;
+
+/// Clock used by [`Scorer`].
+#[cfg(feature = "no-std")]
+pub type DefaultClock = AlwaysPresent;
+
+/// [`routing::Score`] implementation parameterized by a [`Clock`].
+///
+/// See [`Scorer`] for details.
+pub struct ScorerUsingClock<C: Clock + Sub<Duration, Output = C>> {
+	params: ScoringParameters,
+	channel_failures: HashMap<u64, ChannelFailure<C>>,
+}
+
+/// Parameters for configuring [`Scorer`].
+pub struct ScoringParameters {
+	/// A fixed penalty in msats to apply to each channel.
+	pub base_penalty_msat: u64,
+
+	/// A penalty in msats to apply to a channel upon failure.
+	///
+	/// This may be reduced over time based on [`failure_penalty_half_life`].
+	///
+	/// [`failure_penalty_half_life`]: Self::failure_penalty_half_life
+	pub failure_penalty_msat: u64,
+
+	/// The time needed before any accumulated channel failure penalties are cut in half.
+	pub failure_penalty_half_life: Duration,
+}
+
+impl_writeable_tlv_based!(ScoringParameters, {
+	(0, base_penalty_msat, required),
+	(2, failure_penalty_msat, required),
+	(4, failure_penalty_half_life, required),
+});
+
+/// Accounting for penalties from channel failures.
+///
+/// Penalties decay over time, though accumulated as more failures occur.
+struct ChannelFailure<C: Clock + Sub<Duration, Output = C>> {
+	/// Accumulated penalty in msats for the channel as of `last_failed`.
+	undecayed_penalty_msat: u64,
+
+	/// Last time the channel failed. Used to decay `undecayed_penalty_msat`.
+	last_failed: C,
+}
+
+/// A measurement of time.
+pub trait Clock {
+	/// Returns an instance corresponding to "now".
+	fn now() -> Self;
+
+	/// Returns the amount of time elapsed since the clock was created.
+	fn elapsed(&self) -> Duration;
+}
+
+impl<C: Clock + Sub<Duration, Output = C>> ScorerUsingClock<C> {
```

Comment:
Why not just make the `Sub` requirement a direct parent of `Clock`?

Reply:
Good point. Done!